src/haz3lcore/statics/Statics.re

```diff
@@ -260,10 +273,34 @@ and uexp_to_info_map =
     );
   | Let(p, def, body) =>
     let (p_syn, _m) = go_pat(~is_synswitch=true, ~mode=Syn, p, m);
-    let def_ctx = extend_let_def_ctx(ctx, p, p_syn.ctx, def);
-    let (def, m) = go'(~ctx=def_ctx, ~mode=Ana(p_syn.ty), def, m);
-    /* Analyze pattern to incorporate def type into ctx */
-    let (p_ana, m) = go_pat(~is_synswitch=false, ~mode=Ana(def.ty), p, m);
+    let (def, p_ana, m) =
+      if (!is_recursive(ctx, p, def, p_syn.ty)) {
+        let (def, m) = go(~mode=Ana(p_syn.ty), def, m);
+        let (p_ana, m) =
+          go_pat(~is_synswitch=false, ~mode=Ana(def.ty), p, m);
+        (def, p_ana, m);
+      } else {
+        let (def_base, _m) = go'(~ctx=p_syn.ctx, ~mode=Ana(p_syn.ty), def, m) /* Analyze pattern to incorporate def type into ctx */;
```

Comment:
Attempts to synthesize out more accurate types for recursive functions:
e.g.: for unannotated functions `odd` and `even`, 
![image](https://github.com/hazelgrove/hazel/assets/108375845/028d98ba-b641-4907-a3c3-a0353c770a68)
* `p_syn.ctx` has `odd: ?, even: ?`
* `def_base.ty` = `odd: ? -> Bool, even: ? -> ?`